### PR TITLE
Players cannot observe before PREGAME is reached

### LINKED ISSUE #28157: People who observe before the mapping(?) subsystem kicks in aren't initialized
kind of an issue
@Cyberboss

### LINKED ISSUE #28205: If you observe fast enough, lobby music will play in game
This is perturbing

### PR DESCRIPTION
**Pull request by optimumtact**
Fixes #28157 
Fixes #28205

Observing is now a toggle before GAME_STATE_PREGAME is reached, after that it will act as normal and spawn observers.

Observers are now spawned when the game state is set to PREGAME

### CODE REVIEW

File: code/modules/mob/dead/new_player/new_player.dm [hunk 1]
```diff
@@ -116,6 +116,9 @@
 		new_player_panel()
 
 	if(href_list["observe"])
+		if(SSticker.currenT-state < GAME_STATE_PREGAME)
+			to_chat(src, "<span class='warning'>You can't join the game while it's setting up, wait for the intialisations to finish.");
```
Comment (ChangelingRain) [inline code/modules/mob/dead/new_player/new_player.dm]: ;

Comment (optimumtact) [inline code/modules/mob/dead/new_player/new_player.dm]: every time

**Comment by optimumtact** [comment]
@Cyberboss can you confirm if this is post intialisations or do I need to add a new gamestate?

**Comment by optimumtact** [comment]
Also you should really document the game states and their flows

**Comment by optimumtact** [comment]
Ready for review

File: code/modules/mob/dead/new_player/new_player.dm [hunk 2]
```diff
@@ -116,33 +120,14 @@
...
-			observer.update_icon()
-			observer.stop_sound_channel(CHANNEL_LOBBYMUSIC)
-			QDEL_NULL(mind)
-			qdel(src)
-			return 1
+		wants_to_observe = text2num(href_list["observe"])
+		//If we're hot to trot, lets just send them into observing straight away
+		//Otherwise they'll be made observers when game_state changes
+		if(SSticker.current_state < GAME_STATE_PREGAME && wants_to_observe)
+			to_chat(src, "<span class='warning'>You can't join the game while it's setting up, wait for the intialisations to finish.")
```
Comment (optimumtact) [inline code/modules/mob/dead/new_player/new_player.dm]: This warning can probably safely be skipped

**Comment by AnturK** [comment]
Ready should probably be disabled when you sign for observe. You remove new player so it shouldn't matter much but will just look bad.

**Comment by optimumtact** [comment]
yes I am working on that

this is actually WIP again

**Comment by optimumtact** [comment]
Ready for review and tested, it's a full toggle like Ready | not Ready now, and will also toggle ready status.

I thought about combining the three states into one selector Ready | Not Ready | Observer, but I'm afraid of disrupting peoples muscle memory

**Comment by Cyberboss** [comment]
There's really nothing wrong with players observing before initializations. The bug here is that the ghosts are not getting initialized considering I've explicitly coded them to initialize even before the map loads

**Comment by Cyberboss** [comment]
Also, I can't reproduce the bug. @ChangelingRain ?

**Comment by Cyberboss** [comment]
The language thing is a different bug. If you VV yourself, you can see that
initialized is set to 1

**Comment by MollyKristoph** [comment]
The language bit initializes some time around the point where your ghost customizations do, and isn't that big a deal for chronic observers since it's rare that anyone is talking in saltchat anyway before the initializations take place.

File: code/controllers/subsystem/job.dm [hunk 3]
```diff
@@ -224,7 +224,7 @@ SUBSYSTEM_DEF(job)
 
 	//Get the players who are ready
 	for(var/mob/dead/new_player/player in GLOB.player_list)
-		if(player.ready && player.mind && !player.mind.assigned_role)
+		if(player.ready == PLAYER_READY_TO_PLAY && player.mind && !player.mind.assigned_role)
```
Comment (optimumtact) [inline code/controllers/subsystem/job.dm]: anyone know why travis is complaining about this line?

Comment (AnturK) [inline code/controllers/subsystem/job.dm]: You didn't actually set values for the defines. They're just empty tokens now.

Comment (optimumtact) [inline code/controllers/subsystem/job.dm]: Fuck I'm a moron

**Comment by optimumtact** [comment]
it's ready

**Comment by optimumtact** [comment]
I apologise I fell asleep before I could leave an updated comment after testing last night

**Pull request merged**